nucypher/policy/conditions/lingo.py

```diff
@@ -3,68 +3,118 @@
 import json
 import operator as pyoperator
 from hashlib import md5
-from typing import Any, Dict, Iterator, List, Optional, Union
+from typing import Any, List, Optional, Tuple
 
-from marshmallow import fields, post_load
+from marshmallow import fields, post_load, validate
 
 from nucypher.policy.conditions.base import AccessControlCondition
 from nucypher.policy.conditions.context import is_context_variable
 from nucypher.policy.conditions.exceptions import (
-    InvalidConditionLingo,
     InvalidLogicalOperator,
     ReturnValueEvaluationError,
 )
-from nucypher.policy.conditions.types import LingoList, OperatorDict
+from nucypher.policy.conditions.types import Lingo
 from nucypher.policy.conditions.utils import (
     CamelCaseSchema,
     deserialize_condition_lingo,
 )
 
 
-class Operator:
-    OPERATORS = ("and", "or")
+class _ConditionsField(fields.Dict):
+    """Serializes/Deserializes Conditions to/from dictionaries"""
 
-    def __init__(self, _operator: str):
-        if _operator not in self.OPERATORS:
-            raise InvalidLogicalOperator(f"{_operator} is not a valid operator")
-        self.operator = _operator
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-    def __str__(self) -> str:
-        return self.operator
+    def _serialize(self, value, attr, obj, **kwargs):
+        return value.to_dict()
 
-    def to_dict(self) -> OperatorDict:
-        # strict typing of operator value (must be a literal)
-        if self.operator == "and":
-            return {"operator": "and"}
-        else:
-            return {"operator": "or"}
+    def _deserialize(self, value, attr, data, **kwargs):
+        condition = deserialize_condition_lingo(value)
+        return condition
 
-    @classmethod
-    def from_dict(cls, data: OperatorDict) -> "Operator":
-        cls.validate(data)
-        instance = cls(_operator=data["operator"])
-        return instance
 
-    @classmethod
-    def from_json(cls, data) -> 'Operator':
-        data = json.loads(data)
-        instance = cls.from_dict(data)
-        return instance
+#
+# CONDITION = BASE_CONDITION | COMPOUND_CONDITION
+#
+# BASE_CONDITION = {
+#     // ..
+# }
+#
+# COMPOUND_CONDITION = {
+#     "operator": OPERATOR,
+#     "operands": [CONDITION*]
+# }
 
-    def to_json(self) -> str:
-        data = self.to_dict()
-        json_data = json.dumps(data)
-        return json_data
 
-    @classmethod
-    def validate(cls, data: Dict) -> None:
-        try:
-            _operator = data["operator"]
-        except KeyError:
-            raise InvalidLogicalOperator(f"Invalid operator data: {data}")
+class CompoundAccessControlCondition(AccessControlCondition):
```

Comment:
That's a very sensible restriction, in fact I'd make the ceiling much lower, perhaps 5-10 elements since each call adds o(n) runtime at minimum, and maps directly onto node operational costs.  Even beyond that, we may consider some form of rate limiting.  All of which needs to be addressed in another PR IMO.

Reply:
Nice thought!

We can potentially limit the length of the list of conditions, and raise an `InvalidCondition` exception if too long during object creation/schema validation.

I'd rather do that than allow the object to be created then fail during execution because it is too long - could be too confusing to the user. That being said, technically you can work around it by having repeated nested compound conditions each with operands under the limit. 

So maybe some kind of overall count needs to be calculated during creation...?

Some good thoughts here. Let's tackle it in another PR.

Reply:
Just to throw is out there, another way to enforce a limitation is by capping the number of bytes.